### ISSUE TITLE
Mark feature "support OCI Referrer API" as experimental feature

### ISSUE THREAD

**Issue by yizha1** [notaryproject/notation#659]
### What is the areas you experience the issue in?

Notation CLI

### What is not working as expected?

Currently [Referrer API](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers) is by default verified by Notation since `v1.0.0-rc.4`. For example, when using `notation sign`, if registries support referrer API, notation will not use an image index pushed to a tag described by the [referrers tag schema](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#referrers-tag-schema). Since OCI spec 1.1 (including both image spec and distribution spec) is not officially released yet, support of Referrer API could lead to breaking changes or portability issue after Notation v1 is released. On the other hand, users may be confused that sometimes image index is pushed, sometimes not, which is an issue of consistency.

### What did you expect to happen?

Mark feature "support OCI Referrer API" as experimental feature, and document it properly. Users can use experimental features only after env variable `NOTATION_EXPERIMENTAL=1` is set. 
The default behavior of Notation is always using an image index pushed to a tag described by the [referrers tag schema](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc2/spec.md#referrers-tag-schema). If users set env variable `NOTATION_EXPERIMENTAL=1`, Notation will check the support of OCI referrer API first, if registries support Referrer API, then notation will not push image index; if registries don't support Referrer API, notation will push image index as the default behavior. 

### How can we reproduce it?

N/A

### Describe your environment

WSL

### What is the version of your Notation CLI or Notation Library?

Notation `v1.0.0-rc.4`

**Comment by yizha1** [comment]
@toddysm @iamsamirzon @sajayantony @shizhMSFT @vaninrao10 This issue was created per discussion during  community call on 5/8/2023. Please take a look, and this issue should be addressed in Notation v1 release

**Comment by Two-Hearts** [comment]
Thanks @yizha1, I will take this one.

**Comment by Two-Hearts** [comment]
I'm trying to summarize behaviors between Notation versions:
`Sign` as of Notation v1.0.0-rc.4 (current version):
1. If store signature with `image manifest`: Try Referrers API first, if not supported, automatically fallback to referrers tag schema.
2. If store signature with `artifact manifest`: ping Referrers API, if failed (i.e. Referrers API is not supported), then the whole Sign process is failed, no fallback would occur.

`Verify` as of Notation v1.0.0-rc.4 (current version):
Regardless of manifest type, always try Referrers API, if not supported, automatically fallback to referrers tag schema.

What we want in v1.0 for `Sign`:
1. If experimental flag is disabled (NOTATION_EXPERIMENTAL=0), always use referrers tag schema to store signatures.
2. If experimental flag is enabled (NOTATION_EXPERIMENTAL=1), then try referrers API first, if not supported, automatically fallback to referrers tag schema.

What we want in v1.0 for `Verify`:
Q: what is the expected behavior here? @yizha1 /cc: @iamsamirzon @vaninrao10 @sajayantony @toddysm @shizhMSFT @FeynmanZhou

**Comment by shizhMSFT** [comment]
We can assume the usage of `NOTATION_EXPERIMENTAL` is consistent and most customer will not enable this flag.

**Comment by iamsamirzon** [comment]
> 2\. rimental flag is enabled (NOTATION_EXPERIMENTAL=1), then try referrers API first, if not supported, automatically fallback to referrers tag schema.

I like the proposal from @Two-Hearts about having the Notation Verify check for referrers API first and if not available use the image-index method. As long as this does not cause inconsistent results for registries this may be a good path. Essentially be more flexible when verifying signatures so users do not have to change their deployment time checks (modify scripts or flags for verify) when their registries start supporting 1.1 based signatures

**Comment by yizha1** [comment]
> My 2 cents, for v1.0
> 
> ### For Sign:
> * If experimental flag is disabled (NOTATION_EXPERIMENTAL=0), always use referrers tag schema to store signatures. Also, customers should not be able to use `--signature-manifest` flag


The rationale for commands `verify`, `ls` and `inspect` makes sense. But Referrer API doesn't have to be with specific manifest type, that means, no matter what signature manifest is used. We have another issue to remove `--signature-manifest` flag, since artifact manifest has been removed by recent [OCI image spec v1.0.0-rc.3](https://github.com/opencontainers/image-spec/commit/63b8bd02f5b5a2ce464a9a8ea6df049c326ce20f) release.

**Comment by priteshbandi** [comment]
@yizha1 Yes, we should remove `--signature-manifest` flag (in rc5) and introduce a new flag in the future(1.0) using which users can control if they want to push a signature using tag-schema or referrers API.

Rationale:
1. OCI spec removed artifact manifest.
2. For backward compatibility in CI/CD tooling. Think about a scenario where a user is using a managed 3rd party registry like ACR, ECR, docker hub and has CICD tooling around it, a notation to sign an image, and a signature verification tool(either homegrown or 3rd party) that supports only tag schema to verify the image.
As of today, the registry doesn't support referrers API so notation signs with tag schema and verification tool verify image using tag schema. One day, the registry added support for referrers API and suddenly notation started signing using referrers API, which the verification tool doesn't understand, and that broke the user's CICD.

**Comment by Two-Hearts** [comment]
Thanks @priteshbandi, to summarize what we want in RC.5:
For `Sign`:
1. `--signautre-manifest` flag would be removed, we will not support signing with OCI artifact manifest anymore.
2. Adding a new flag, `--use-referrers-api` (name TBD), behind `experimental`:
      - When both `NOTATION_EXPERIMENTAL` and `--use-referrers-api` are set, Notation would try the Referrers API first, if failed, automatically fallback to referrers tag schema.
      - Otherwise, `Sign` will be using referrers tag schema as default.

For other commands that consume signatures, `notation list`, `notation verify`, etc.:
1. Notation still supports OCI artifact manifest.
2. Always try the Referrers API first, if failed, automatically fallback to the referrers tag schema.

**Comment by shizhMSFT** [comment]
> Always try the Referrers API first, if failed, automatically fallback to the referrers tag schema.

There might be a compatibility issue with some registries such as GAR.

Basically, attempting the Referrers API to GAR will not return a 404 but an unexpected 302.

See discussions:
- https://cloud-native.slack.com/archives/CQUH8U287/p1681159105868649
- #640

**Comment by priteshbandi** [comment]
@Two-Hearts 
> For Sign:
Adding a new flag, --use-referrers-api (name TBD), behind experimental:
    When both NOTATION_EXPERIMENTAL and --use-referrers-api are set, Notation would try the Referrers API first, if failed, automatically fallback to referrers tag schema.
    Otherwise, Sign will be using referrers tag schema as default

IMO we should not fallback for sign operation i.e. When both NOTATION_EXPERIMENTAL and `--use-referrers-api` are set, Notation would try the Referrers API only. If failed, fail the sign operation and ask user to not use  `--use-referrers-api`  flag.

@shizhMSFT 
IMO automatic fallback would useful for most of the registries in OCI 1.0 to 1.1 migration usecase. For scenarios where registries are misbehaving not abiding by [spec](https://github.com/opencontainers/distribution-spec/blob/57b69b6474e2299926802bb10014074d8592b6b8/spec.md#unavailable-referrers-api) we can have --not-use-referrers-api(name TBD) flag.

**Comment by Two-Hearts** [comment]
> IMO we should not fallback for sign operation i.e. When both NOTATION_EXPERIMENTAL and `--use-referrers-api` are set, Notation would try the Referrers API only. If failed, fail the sign operation and ask user to not use `--use-referrers-api` flag.

Yeah, this also works. In this way, the `Sign` command becomes more deterministic (without any automatic fallback). Waiting for others' comments on this one.

**Comment by Two-Hearts** [comment]
> > Always try the Referrers API first, if failed, automatically fallback to the referrers tag schema.
> 
> There might be a compatibility issue with some registries such as GAR.
> 
> Basically, attempting the Referrers API to GAR will not return a 404 but an unexpected 302.
> 
> See discussions:
> 
> * https://cloud-native.slack.com/archives/CQUH8U287/p1681159105868649
> * [Introduce an experimental flag to enable backward compatibility with OCI registries #630](https://github.com/notaryproject/notation/issues/630)

@shizhMSFT hmm, yes, we need to resolve issue [630](https://github.com/notaryproject/notation/issues/630). The question is do we want to have it done in RC.5 or v1.0?

**Comment by yizha1** [comment]
> > IMO we should not fallback for sign operation i.e. When both NOTATION_EXPERIMENTAL and `--use-referrers-api` are set, Notation would try the Referrers API only. If failed, fail the sign operation and ask user to not use `--use-referrers-api` flag.
> 
> Yeah, this also works. In this way, the `Sign` command becomes more deterministic (without any automatic fallback). Waiting for others' comments on this one.

Yes, it makes sense. Use the experimental flag means you want to leverage Referrer API, and not tag schema. Otherwise, it may succeed with tag schema, which is not what users expected.

**Comment by yizha1** [comment]
> There might be a compatibility issue with some registries such as GAR.
> 
> Basically, attempting the Referrers API to GAR will not return a 404 but an unexpected 302.
> 
> See discussions:
> 
> * https://cloud-native.slack.com/archives/CQUH8U287/p1681159105868649
> * [Introduce an experimental flag to enable backward compatibility with OCI registries #630](https://github.com/notaryproject/notation/issues/630)

Maye we should raise it to OCI spec maintainers on the determination of referrer API

**Comment by shizhMSFT** [comment]
In order to make `notation` stable while we are building `notation` on top of the instable `distribution-spec`, we need to identify which part of the `distribution-spec` is relatively stable and which is not. Apparently, the referrers tag schema is relatively stable and should be used as the default behavior of `notation`. Meanwhile, the referrers API and its fallback strategy is instable and should be marked as _experimental_. 

Based on the above fundamentals, I'd like to propose a new _experimental_ flag `--allow-referrers-api` to all registry related commands such as `notation sign` and `notation verify`.

- When `--allow-referrers-api` is toggled with `NOTATION_EXPERIMENTAL=1`, `notation` attempts the referrers API and fallback to referrers tag schema on failure.
- Otherwise, `notation` has deterministic behavior and uses referrers tag schema.

From previous discussions, we have consensus on using referrers tag schema by default for `notation sign` but not aligned on `notation verify`. To me, `notation verify` should not attempt referrers API without `NOTATION_EXPERIMENTAL=1` since the referrers API itself is instable / experimental.

**Comment by Two-Hearts** [comment]
> In order to make `notation` stable while we are building `notation` on top of the instable `distribution-spec`, we need to identify which part of the `distribution-spec` is relatively stable and which is not. Apparently, the referrers tag schema is relatively stable and should be used as the default behavior of `notation`. Meanwhile, the referrers API and its fallback strategy is instable and should be marked as _experimental_.
> 
> Based on the above fundamentals, I'd like to propose a new _experimental_ flag `--allow-referrers-api` to all registry related commands such as `notation sign` and `notation verify`.
> 
> * When `--allow-referrers-api` is toggled with `NOTATION_EXPERIMENTAL=1`, `notation` attempts the referrers API and fallback to referrers tag schema on failure.
> * Otherwise, `notation` has deterministic behavior and uses referrers tag schema.
> 
> From previous discussions, we have consensus on using referrers tag schema by default for `notation sign` but not aligned on `notation verify`. To me, `notation verify` should not attempt referrers API without `NOTATION_EXPERIMENTAL=1` since the referrers API itself is instable / experimental.

Yeah, after discussion with Shiwei offline, I agree with the above design. The `automatic fallback from referrers API to tag schema` is a part of the Referrers API's [spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers). As long as a user attempts the Referrers API, the process should include the fallback mechanism to align with the spec. From this point of view, I think we can still claim that Notation is deterministic regarding the use of Referrers API and Referrers tag schema.
@priteshbandi What do you think?

**Comment by priteshbandi** [comment]
I agree, to summarize
### If experimental flag is disabled (NOTATION_EXPERIMENTAL=0), dont use referrers API i.e.
* For Sign, only use referrers tag schema to store signatures.
* For verify,inspect, list, etc, only use referrers tag schema to pull signatures.



###  If experimental flag is enabled (NOTATION_EXPERIMENTAL=1) and user uses `--use-referrers-api` flag; use oci 1.1 spec i.e. use referrers API
* For Sign, Check if referrers API returns 2xx or not. If it doesnt returns 2xx fail signing because registry hasnt correctly/fully implemented oci1.1. If it return 2xx then sign using referrers.
* For verify,inspect, list, etc use referrers API first to pull the signature.
  * If referrers API returns 2xx use it to list signatures
  * If referrers API returns 404, fallback and use tag schema to list signatures
  * If referrers API returns anything else, fail the operation

**Comment by Two-Hearts** [comment]
> #### If experimental flag is enabled (NOTATION_EXPERIMENTAL=1) and user uses `--use-referrers` flag; use oci 1.1 spec i.e. use referrers API
> * For sign, Check if referrers API returns 2xx or not. If it doesnt returns 2xx fail signing because registry hasnt correctly/fully implemented oci1.1. If it return 2xx then sign using referrers.

Thanks @priteshbandi, everything else LGTM. My only question is, when NOTATION_EXPERIMENTAL=1 and uses `--allow-referrers` flag during `Sign`, should we still enable fallback to referrers tag schema if referrers API is not supported? Because the fallback mechanism is a part of the Referrers API spec, trying referrers API without enabling the fallback schema seems not strictly following the distribution spec?
Quoting from the spec: `If the [referrers API](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers) returns a 404, the client MUST fallback to pulling the [referrers tag schema](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#referrers-tag-schema).`
Reference: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers
/cc: @shizhMSFT

**Comment by priteshbandi** [comment]
My interpretation of spec is that (correct me if I am wrong as I am not closely following spec) fallback is only for pull operation but since sign is an push operation its better to provide consistency . Also, if user is opting in for oci spec 1.1 we should always use referrers API related behaviors. Want to hear others opinion.

**Comment by yizha1** [comment]
> My interpretation (I am not closely following spec) of spec is that fallback is only for pull operation but since sign is an push operation its better to provide consistency . Also, if user is opting in for oci spec 1.1 we should always use referrers API related behaviors. Want to hear others opinion.

@priteshbandi The push manifest procedure is defined here https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc2/spec.md#pushing-manifests-with-subject

**Comment by priteshbandi** [comment]
Thanks @Two-Hearts and @yizha1  for pointing to the right spec. As oci1.1-rc2 calls out, when user adds NOTATION_EXPERIMENTAL=1 and uses --allow-referrers flag, notation should perform fallback for sign operation.
